desc/perturbations.py

```diff
@@ -610,38 +574,29 @@ def optimal_perturb(  # noqa: C901 - FIXME: break this up into simpler pieces
     dx2_reduced = 0
 
     # dx/dx_reduced
-    dxdx_reduced = jnp.eye(objective_f.dim_x)[:, unfixed_idx] @ Z
+    dxdx_reduced = jnp.eye(eq.dim_x)[:, unfixed_idx] @ Z
 
     # dx/dc
-    dxdc = jnp.zeros((objective_f.dim_x, 0))
-    if len(
-        [
-            arg
-            for arg in ("R_lmn", "Z_lmn", "L_lmn", "p_l", "i_l", "Psi")
-            if arg in deltas.keys()
-        ]
-    ):
-        x_idx = jnp.concatenate(
-            [objective_f.x_idx[arg] for arg in arg_order if arg in deltas.keys()]
-        )
-        x_idx = jnp.sort(x_idx)
-        dxdc = jnp.eye(objective_f.dim_x)[:, x_idx]
-    if "Rb_lmn" in deltas.keys():
-        con = get_instance(constraints, BoundaryRSelfConsistency)
-        A = con.derivatives["jac_unscaled"]["R_lmn"](
-            *[jnp.zeros(con.dimensions[arg]) for arg in con.args]
-        )
-        Ainv = jnp.linalg.pinv(A)
-        dxdRb = jnp.eye(objective_f.dim_x)[:, objective_f.x_idx["R_lmn"]] @ Ainv
-        dxdc = jnp.hstack((dxdc, dxdRb))
-    if "Zb_lmn" in deltas.keys():
-        con = get_instance(constraints, BoundaryZSelfConsistency)
-        A = con.derivatives["jac_unscaled"]["Z_lmn"](
-            *[jnp.zeros(con.dimensions[arg]) for arg in con.args]
-        )
-        Ainv = jnp.linalg.pinv(A)
-        dxdZb = jnp.eye(objective_f.dim_x)[:, objective_f.x_idx["Z_lmn"]] @ Ainv
-        dxdc = jnp.hstack((dxdc, dxdZb))
+    dxdc = []
+    xz = objective_f.unpack_state(jnp.zeros_like(xf), False)[0]
+
+    for arg in deltas:
+        if arg not in ["Rb_lmn", "Zb_lmn"]:
+            x_idx = eq.x_idx[arg]
+            dxdc.append(jnp.eye(eq.dim_x)[:, x_idx])
+        if arg == "Rb_lmn":
+            con = get_instance(constraints, BoundaryRSelfConsistency)
+            A = con.jac_unscaled(xz)["R_lmn"]
```

Comment:
So basically, you get the derivative df/dx and the constraints function applies the constraints somehow.
How does this work?